arrows/ffmpeg/ffmpeg_video_output.cxx

```diff
@@ -0,0 +1,498 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/// \file
+/// \brief Implementation of FFmpeg video writer.
+
+#include "arrows/ffmpeg/ffmpeg_init.h"
+#include "arrows/ffmpeg/ffmpeg_video_output.h"
+#include "arrows/ffmpeg/ffmpeg_video_settings.h"
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/imgutils.h>
+#include <libswscale/swscale.h>
+}
+
+namespace kv = kwiver::vital;
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace ffmpeg {
+
+class ffmpeg_video_output::impl
+{
+public:
+  impl();
+
+  ~impl();
+
+  bool write_next_packet();
+
+  void write_remaining_packets();
+
+private:
+  friend class ffmpeg_video_output;
+
+  AVFormatContext* format_context;
+  AVOutputFormat* output_format;
+  AVStream* video_stream;
+  AVStream* metadata_stream;
+  AVCodecContext* codec_context;
+  AVCodec* codec;
+  SwsContext* image_conversion_context;
+
+  kv::logger_handle_t logger;
+
+  size_t frame_count;
+  AVRational frame_rate;
+};
+
+ffmpeg_video_output::impl
+::impl()
+  : format_context{ nullptr },
+    output_format{ nullptr },
+    video_stream{ nullptr },
+    metadata_stream{ nullptr },
+    codec_context{ nullptr },
+    codec{ nullptr },
+    image_conversion_context{ nullptr },
+    logger{},
+    frame_count{ 0 },
+    frame_rate{ 0, 1 }
+{
+  ffmpeg_init();
+}
+
+ffmpeg_video_output::impl::~impl() {}
+
+bool
+ffmpeg_video_output::impl
+::write_next_packet()
+{
+  AVPacket packet = {};
+  av_init_packet( &packet );
+
+  // Attempt to read next encoded packet
+  auto const success =
+    avcodec_receive_packet( codec_context, &packet );
+
+  if( success == AVERROR( EAGAIN ) || success == AVERROR_EOF )
+  {
+    // Failed expectedly: no packet to read
+    av_packet_unref( &packet );
+    return false;
+  }
+  if( success < 0 )
+  {
+    // Failed unexpectedly
+    av_packet_unref( &packet );
+    VITAL_THROW( kv::video_runtime_exception, "Failed to receive packet" );
+  }
+
+  // Succeeded; write to file
+  if( av_interleaved_write_frame( format_context, &packet ) < 0 )
+  {
+    VITAL_THROW( kv::video_runtime_exception, "Failed to write packet" );
+  }
+
+  return true;
+}
+
+void
+ffmpeg_video_output::impl
+::write_remaining_packets()
+{
+  // Enter "draining mode" - i.e. signal end of file
+  avcodec_send_frame( codec_context, nullptr );
+
+  while( write_next_packet() ) {}
+}
+
+ffmpeg_video_output
+::ffmpeg_video_output() : d{ new impl{} } {
```

Comment:
Should the final `{` on this line be on the next line by itself?

Reply:
Good catch! I must have forgotten to re-run uncrustify. Fixed.